lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1058,6 +1058,36 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
     }
 
     pattern = newPattern;
+
+    // Use of 'as!' or 'as?' will cause errors in the type checking below.
+    // Catch that here and give a clear error message.
+    if (auto VarPat = dyn_cast<VarPattern>(pattern)) {
+      if (auto ExprPat = dyn_cast<ExprPattern>(VarPat->getSubPattern())) {
+        if (auto SubExpr = ExprPat->getSubExpr())
+          if (auto SeqExpr = dyn_cast<SequenceExpr>(SubExpr)) {
+            SourceLoc LastPunctLoc;
+            SeqExpr->forEachImmediateChildExpr([&](Expr *E) -> Expr * {
+              SourceLoc PunctLoc;
+              bool IsForce;
+              if (auto Asq = dyn_cast<ConditionalCheckedCastExpr>(E)) {
+                PunctLoc = Asq->getQuestionLoc();
+                IsForce = false;
+              } else if (auto Asf = dyn_cast<ForcedCheckedCastExpr>(E)) {
+                PunctLoc = Asf->getExclaimLoc();
+                IsForce = true;
+              } else
+                return E; // This child expr isn't one we're looking for.
+
+              if (PunctLoc != LastPunctLoc) // Only 1 diag per `as?`
+                TC.diagnose(E->getLoc(), diag::case_as_punct, IsForce)
+                    .fixItRemove(SourceRange(PunctLoc));
+              LastPunctLoc = PunctLoc;
+              return new (TC.Context) ErrorExpr(E->getSourceRange());
```

Comment:
The return value's actually ignored, so I decided to remove it in #27947. Thanks for the prompt to do so…and sorry for breaking your PR again.

Reply:
I was wrong about the return value being ignored! And I guess this is what keeps the other diagnostic from being emitted?